### PR TITLE
Add Untrusted Deserialization vulnerability

### PR DESCRIPTION
**Pull request by Mariovido**
# What Does This Do
Add Untrusted Deserialization vulnerability

# Motivation
Add a new vulnerability

# Additional Notes

Jira ticket: [APPSEC-17165]

[APPSEC-17165]: https://datadoghq.atlassian.net/browse/APPSEC-17165?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/sink/UntrustedDeserializationModuleTest.groovy [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    when:
+    taint(inputStream)
+    module.onInputStream(inputStream)
+
+    then: 'with tainted input stream'
+    1 * reporter.report(_, { Vulnerability vul -> vul.type == VulnerabilityType.UNTRUSTED_DESERIALIZATION})
+  }
+
+  private void taint(final Object value) {
+    ctx.getTaintedObjects().taint(value, Ranges.forObject(new Source(SourceTypes.REQUEST_BODY, 'name', value.toString())))
```
Comment (jandro996) [inline dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/sink/UntrustedDeserializationModuleTest.groovy:52]:
```suggestion
    ctx.getTaintedObjects().taint(value, [] as Range[])
```

**Review by jandro996 — APPROVED** [review]
LGTM

**Review by manuel-alvarez-alvarez — APPROVED** [review]
LGTM

I think we can expand the smoke testing in a separate PR to ensure we are not losing propagation in different scenarios:

1. Other smoke tests than spring boot as we might be losing propagation of the body (e.g. vertx, jetty, ...)
2. Multipart uploads with multiple binaries (using servlet or other APIs like commons file upload)

They are not directly related to untrusted deserialization but they really influence the rule :sweat: 

PS: Nice first PR!

File: dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/ObjectInputStreamCallSite.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import datadog.trace.api.iast.IastCallSites;
+import datadog.trace.api.iast.InstrumentationBridge;
+import datadog.trace.api.iast.Sink;
+import datadog.trace.api.iast.VulnerabilityTypes;
+import datadog.trace.api.iast.sink.UntrustedDeserializationModule;
+import java.io.InputStream;
+
+@Sink(VulnerabilityTypes.UNTRUSTED_DESERIALIZATION)
+@CallSite(spi = IastCallSites.class)
+public class ObjectInputStreamCallSite {
```
Comment (anderruiz) [inline dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/ObjectInputStreamCallSite.java:13]: Do we plan to fix this later? technically the vulnerability happens when readObject is executed. Shouldn't we follow the same pattern that @manuel-alvarez-alvarez started to taint the input stream here but then have the vulnerability when it is loaded. I am also worried about the posible FP that we may have, we were previously only launching the vulnerability on readObject and whenever a banned class was loaded or a process execution was reached

File: dd-smoke-tests/iast-util/src/main/java/datadog/smoketest/springboot/controller/IastWebController.java [hunk 3]
```diff
@@ -394,6 +395,12 @@ public String headerInjectionRedaction(
     return "Ok";
   }
 
+  @GetMapping("/untrusted_deserialization")
+  public String untrustedDeserialization(HttpServletRequest request) throws IOException {
+    ObjectInputStream ois = new ObjectInputStream(request.getInputStream());
```
Comment (anderruiz) [inline dd-smoke-tests/iast-util/src/main/java/datadog/smoketest/springboot/controller/IastWebController.java:400]: This for example should not cause a UD vulnerabilities

**Review by anderruiz — COMMENTED** [review]
Left some comments

File: dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/ObjectInputStreamCallSite.java (hunk 2 above)
Comment (Mariovido) [reply dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/ObjectInputStreamCallSite.java:13]: We have discussed it internally and have decided to follow the current approach and address it in another independent PR

File: dd-smoke-tests/iast-util/src/main/java/datadog/smoketest/springboot/controller/IastWebController.java (hunk 3 above)
Comment (Mariovido) [reply dd-smoke-tests/iast-util/src/main/java/datadog/smoketest/springboot/controller/IastWebController.java:400]: Same as above :)

**Pull request merged**